### ISSUE TITLE
curl_easy_perform hangs with imap and PolarSSL

### ISSUE THREAD

**Issue by r-ku** [bagder/curl#334]
Code:

```c
    CURL* curl = curl_easy_init();

    curl_easy_setopt(curl, CURLOPT_USERNAME, "<removed due to obvious reasons>");
    curl_easy_setopt(curl, CURLOPT_PASSWORD, "<removed due to obvious reasons>");
    curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0L);

    curl_easy_setopt(curl, CURLOPT_URL, "imaps://imap.mail.ru:993");
    curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "SELECT /INBOX");

    curl_easy_perform(curl);

    curl_easy_setopt(curl, CURLOPT_URL, "imaps://imap.mail.ru:993/INBOX/;UID=298/;SECTION=TEXT");
    curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, 0);

    curl_easy_perform(curl);    // <-- never returns
```

```
* Rebuilt URL to: imaps://imap.mail.ru:993/
*   Trying 217.69.139.90...
* Connected to imap.mail.ru (217.69.139.90) port 993 (#0)
* PolarSSL: Connecting to imap.mail.ru:993
* ALPN, offering http/1.1
* PolarSSL: Handshake complete, cipher is TLS-RSA-WITH-AES-256-GCM-SHA384
* Dumping cert info:
* cert. version     : 3
* serial number     : 4F:EC:EB:5A:8B:7D:EE:7E:80:B4:AE:AD:BE:E7:3B:95
* issuer name       : C=US, O=GeoTrust Inc., CN=GeoTrust SSL CA - G2
* subject name      : C=RU, ST=RUSSIAN FEDERATION, L=Moscow, O=LLC Mail.ru, OU=IT, CN=*.mail.ru
* issued  on        : 2014-09-18 00:00:00
* expires on        : 2015-09-18 23:59:59
* signed using      : RSA with SHA1
* RSA key size      : 2048 bits
* basic constraints : CA=false
* subject alt name  : *.mail.ru
* key usage         : Digital Signature, Key Encipherment
* ext key usage     : TLS Web Server Authentication, TLS Web Client Authentication

* ALPN, server did not agree to a protocol
* SSL connected
< * OK Welcome 
> A001 CAPABILITY 
< * CAPABILITY IMAP4rev1 ID XLIST UIDPLUS UNSELECT MOVE AUTH=PLAIN AUTH=XOAUTH2 
< A001 OK CAPABILITY completed 
> A002 AUTHENTICATE PLAIN 
< +  
> <removed due to obvious reasons>
< * CAPABILITY IMAP4rev1 ID XLIST UIDPLUS UNSELECT MOVE 
< A002 OK Authentication successful 
> A003 SELECT /INBOX 
< * FLAGS (\Answered \Flagged \Deleted \Draft \Seen) 
* FLAGS (\Answered \Flagged \Deleted \Draft \Seen) 
< * 302 EXISTS 
< * 0 RECENT 
* 302 EXISTS 
< * OK [UIDVALIDITY 1398348242] 
* 0 RECENT 
< * OK [PERMANENTFLAGS (\Answered \Flagged \Deleted \Seen \Draft)] 
* OK [UIDVALIDITY 1398348242] 
< * OK [UIDNEXT 573] 
* OK [PERMANENTFLAGS (\Answered \Flagged \Deleted \Seen \Draft)] 
< A003 OK [READ-WRITE] SELECT completed 
* OK [UIDNEXT 573] 
* Connection #0 to host imap.mail.ru left intact
* Found bundle for host imap.mail.ru: 0x85ba00
* Re-using existing connection! (#0) with host imap.mail.ru
* Connected to imap.mail.ru (217.69.139.90) port 993 (#0)
> A004 SELECT INBOX 
< * FLAGS (\Answered \Flagged \Deleted \Draft \Seen) 
< * 302 EXISTS 
< * 0 RECENT 
< * OK [UIDVALIDITY 1398348242] 
< * OK [PERMANENTFLAGS (\Answered \Flagged \Deleted \Seen \Draft)] 
< * OK [UIDNEXT 573] 
< A004 OK [READ-WRITE] SELECT completed 
> A005 FETCH 298 BODY[TEXT] 
< * 298 FETCH (BODY[TEXT] {116856} 
* Found 116856 bytes to download
--b1_c805a120dd4138fc73a75449d46fa932 
Content-Type: multipart/alternative; 
    boundary="b2_c805a120dd4138fc73a75449d46fa932" 
 
--b2_c805a120dd4138fc73a75449d46fa932 
Content-Type: text/plain; charset=utf-8 
Content-Transfer-Encoding: 8bit 
 
Dear <...lots of ommited email text...>
--b1_c805a120dd4138fc73a75449d46fa932-- 
 
```

imap code exits [here](https://github.com/bagder/curl/blob/master/lib/multi.c#L1411) and then im lost, ``multi_runsingle`` constantly loops in ``CURLM_STATE_DO_DONE`` state seemingly waiting for more data which there is no more. Setting callback it properly receives all 116856 bytes. This does not happen with smaller emails.

**Comment by jay** [comment]
Do you really need the `SELECT /INBOX` before `imaps://imap.mail.ru:993/INBOX/;UID=298/;SECTION=TEXT`? The latter should cause a `SELECT INBOX` before it retrieves the message (and according to your output it does). Try to reproduce using the latest version of the curl tool:
~~~
curl -u user:pass "imaps://imap.mail.ru:993/INBOX/;UID=298/;SECTION=TEXT"
~~~
Please include the version information `curl -V` in your reply if you can reproduce. If you cannot then try to reproduce using the latest libcurl.

**Comment by captain-caveman2k** [comment]
Further to Jay's email and suggestion you shouldn't need to perform the SELECT as a custom command and as such I would recommend removing this as well.

In addition to this I was also wondering:

* Are those small emails less that 16k in size or have you had success with some in between 16k and this particular large email?
* Do you get the same problem if you don't use TLS?
* Have you tried putting a break point in imap_done() - does this function get called over and over again and what happens at line 1495 and the subsequent lines where it should set the state to IMAP_FETCH_FINAL?

PS. Even though you didn't include your CURLOPT_USERNAME and CURLOPT_PASSWORD values you did include the base64 encode authentication string as you used PLAIN authentication :(

Kind Regards

Steve

**Comment by r-ku** [comment]
Thank you for removing that auth string, that was epic fail of me :|

I indeed forgot to include version info, but yes, i tried latest (v7.43.0) right after slightly older version exposed this bug to me.

Things i tried so far:

1. Removing ``SELECT`` command changes nothing at all. Its neither needed nor at fault here.
2. Retrieving smaller email (5148 bytes) works fine, libcurl exits loop just fine then.
3. Using curl binary retrieving big email **works**. It exits as expected.
4. Tried not using TLS only to find out that mail provider does not even support it.
5. Tried building curl binary against both openssl and polarssl - binary always exits as expected.
5. ``imap_done`` gets called just once, and it indeed sets ``IMAP_FETCH_FINAL`` state. However call to ``imap_block_statemach()`` never returns.

**Comment by jay** [comment]
> Using curl binary retrieving big email works. It exits as expected.

Thank you for testing but you didn't post your `curl -V` output, please do when you reply. I've been unable to reproduce what you describe using libcurl. Are you sure your program is using the latest libcurl? If you are using libcurl shared library some different version could be loaded. Add this to see which version your program is actually using:
~~~
  printf("\n\n\nYou are using libcurl/%s\n\n\n", curl_version_info(CURLVERSION_NOW)->version);
~~~
I used example [ShowGmailFolders.c](https://gist.github.com/jay/a43724c98220de85d020) and here's the debug info from a similar sized message in gmail:
~~~
> A004 FETCH 13 BODY[TEXT]
* IMAP 0x8f1640 state change from SELECT to FETCH
< * 13 FETCH (BODY[TEXT] {160776}
* Curl_pp_readresp_ 481 bytes of trailing server response left
* Found 160776 bytes to download
* Written 481 bytes, 160295 bytes are left for transfer
* IMAP 0x8f1640 state change from FETCH to STOP
* DO phase is complete
* STATE: DOING => DO_DONE handle 0x8df668; line 1413 (connection #0)
* STATE: DO_DONE => WAITPERFORM handle 0x8df668; line 1464 (connection #0)
* STATE: WAITPERFORM => PERFORM handle 0x8df668; line 1474 (connection #0)
* readwrite_data: we're done!
* nread <= 0, server closed connection, bailing
* STATE: PERFORM => DONE handle 0x8df668; line 1632 (connection #0)
* Curl_done
* IMAP 0x8f1640 state change from STOP to FETCH_FINAL
< )
< A004 OK Success
* IMAP 0x8f1640 state change from FETCH_FINAL to STOP
* Connection #0 to host imap.gmail.com left intact
~~~

Try that example. Replace the url with your own but leave everything else the same and see what happens.

**Comment by r-ku** [comment]
I got around to compiling curl executable myself and guess what - it hangs. Failing curl -V:

    curl 7.43.0 (Linux) libcurl/7.43.0 mbedTLS/1.3.11 zlib/1.2.8 libidn/1.30
    Protocols: gopher http https imap imaps pop3 pop3s rtsp smtp smtps 
    Features: AsynchDNS IDN IPv6 SSL libz UnixSockets

Working curl -V (from archlinux repos):

    curl 7.43.0 (x86_64-unknown-linux-gnu) libcurl/7.43.0 OpenSSL/1.0.2c zlib/1.2.8 libidn/1.30 libssh2/1.6.0
    Protocols: dict file ftp ftps gopher http https imap imaps pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp 
    Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP UnixSockets

I further tested ShowGmailFolders.c only changing it to use my mail - it indeed hangs.

And i think i narrowed it down: my custom build with openssl works, my custom build with mbedtls hangs. So there must be bug somewhere there. Good news is its not fault of cmake. curl -V of my own build that worked:

    curl 7.43.0 (Linux) libcurl/7.43.0 OpenSSL/1.0.2c zlib/1.2.8 libidn/1.30
    Protocols: gopher http https imap imaps pop3 pop3s rtsp smtp smtps 
    Features: AsynchDNS IDN IPv6 SSL libz UnixSockets

**Comment by jay** [comment]
Thanks for that extra information. I had to do a double take when I saw mbedTLS. I see that's the new name for PolarSSL now that it's part of ARM.

ShowGmailFolders hangs with OpenSSL or PolarSSL?

Are the results always the same for each SSL backend? Can you check what happens if you loop it several times? I'd like to eliminate a server issue clouding your results. This should break on success or try 30 times:
~~~
for (( c=1; c<=30; c++ )); do if curl -v -u user:pass "imaps://imap.mail.ru:993/INBOX/;UID=298/;SECTION=TEXT"; then break; fi; printf "\n"; sleep 3; done
~~~

**Comment by r-ku** [comment]
ShowGmailFolders hangs with PolarSSL. So does ``curl``. However it works as expected with OpenSSL.

I tried loop testcase - my own copy compiled with OpenSSL as well as distro-provided curl (also with OpenSSL) loop fine 30 times. My own compiled curl with PolarSSL hangs on first invocation after printing text.

P.S. Sorry for long delay. Some things came up.

**Comment by bagder** [comment]
Can you help us pinpoint where exactly libcurl hangs on polarssl in this case?

**Comment by r-ku** [comment]
As i already wrote in very first post:

imap code exits [here](https://github.com/bagder/curl/blob/master/lib/multi.c#L1411) and then im lost, ``multi_runsingle`` constantly loops in ``CURLM_STATE_DO_DONE`` state seemingly waiting for more data which there is no more. Setting callback it properly receives all 116856 bytes. This does not happen with smaller emails.

I am not sure how to track it further, but i would be happy to help if i knew what to do.

**Comment by bagder** [comment]
loops in CURLM_STATE_DO_DONE? How can it even get stuck in that state? As I read the code, that state is highly transitional and will immediately move on to either CURLM_STATE_WAITPERFORM or CURLM_STATE_DONE. If it waits for more data it sounds more like it is in the CURLM_STATE_PERFORM state...

This is indeed a very tricky problem since we've failed to repeat it and it seems to require some rather special circumstances to trigger.

**Comment by r-ku** [comment]
Maybe i should post source package + configuration + build scripts + PolarSSL package i used? I have it all rigged to build using cmake although it is a bit of hack. I tried building using autotools by the way - result was the same.

**Comment by bagder** [comment]
PolarSSL most likely needs a fix similar to c111178bd4
